builder/hspkg-builder.nix

```diff
@@ -29,7 +29,13 @@ let
       "ghc902/stm-2.5.0.0" = "/libraries/stm";
       "ghc902/filepath-1.4.2.1" = "/libraries/filepath";
     }."${compiler-nix-name}/${name}" or null;
-  src = if bundledSrc == null then pkg.src else ghc.configured-src + bundledSrc;
+  baseUrlMatch = __match "(.*)/package/([^/]*)" pkg.src.url;
+  src =
+    if bundledSrc != null
+      then ghc.configured-src + bundledSrc
+    else if pkg.src ? url && baseUrlMatch != null && inputMap ? ${__head baseUrlMatch}
+      then inputMap.${__head baseUrlMatch} + "/package/${__elemAt baseUrlMatch 1}"
```

Comment:
This feels a bit evil. We're kind of breaking the abstraction of `inputsMap` as just a map of replacements for specific URLs... hence why I wonder if we can just cajole cabal into looking in the right place to begin with.

Reply:
It is still doing a like for like replacement though (extending the mapping to include a subdirectory).

The pkg.src.url is mapped:

```
from:             "${__head baseUrlMatch}/package/${__elemAt baseUrlMatch 1}"
to:   inputMap.${__head baseUrlMatch} + "/package/${__elemAt baseUrlMatch 1}"
```

> can just cajole cabal into looking in the right place to begin with

One advantage of this approach is that we can run `cabal configure` on an unmodified `cabal.project` file.  We still have to parse `cabal.project` to find out what repositories to download into the `$HOME/.cabal` directory, but unlike the `source-repository-package` blocks we do not need to make changes to `cabal.project`.

Reply:
Or putting it another way, now `inputMap` means two things:
1. prefetched values of the given URLs
2. some kind of url transformer thing that gets applied when fetching package sources

Maybe the thing I'm trying to do is just bad and I should avoid it :thinking: